### PR TITLE
frontend: convert input to typescript

### PR DESCRIPTION
**Pull request by thisconnect**

### CODE REVIEW

**Comment by KasparEtter** [comment]
CI is not yet happy… 🙁

File: frontends/web/src/components/forms/input.tsx [hunk 1]
```diff
@@ -14,27 +14,38 @@
...
 
-import { h } from 'preact';
-import { input, errorText, isTransparent } from './input.css';
+import { h, RenderableProps } from 'preact';
+import * as styles from './input.css';
+
+export interface Props {
+    children?: JSX.Element;
+    className?: string;
+    error: JSON;
```
Comment (KasparEtter) [inline frontends/web/src/components/forms/input.tsx]: I guess `error` is optional as well? This would solve quite a few of your CI issues…

**Comment by KasparEtter** [comment]
Once you've resolved the CI issues, we can merge this PR! 🙂

**Comment by thisconnect** [comment]
not yet, sorry

screen shot 2018-11-22 at 9 28 31 pm

**Comment by thisconnect** [comment]
ci is very happy now, but I still want to test everything a bit more

File: frontends/web/package.json [hunk 2]
```diff
@@ -18,18 +18,18 @@
     "build/*"
   ],
   "devDependencies": {
-    "@types/jest": "23.3.2",
+    "@types/jest": "^23.3.10",
```
Comment (KasparEtter) [inline frontends/web/package.json]: Why do we introduce now a `^`?

Comment (thisconnect) [reply frontends/web/package.json]: good catch, because I didn't check.

`yarn upgrade @types/jest ts-jest --latest`

**Comment by benma** [comment]
@KasparEtter I believe this is awaiting your review :)

File: frontends/web/src/components/forms/input.tsx [hunk 3]
```diff
@@ -14,27 +14,49 @@
...
 
-import { h } from 'preact';
-import { input, errorText, isTransparent } from './input.css';
+import { h, RenderableProps } from 'preact';
+import * as styles from './input.css';
+
+export interface Props {
+    autoComplete?: boolean | 'on' | 'off';
+    autoFocus?: boolean;
+    children?: any; // can be removed once preact > 8.3.1
```
Comment (KasparEtter) [inline frontends/web/src/components/forms/input.tsx]: preact is > 8.3.1 with your PR.

File: frontends/web/test/components/forms/input.test.tsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import { h } from 'preact';
+import { deep, shallow } from 'preact-render-spy';
+
+import Input, { Props } from '../../../src/components/forms/input';
+
+describe('components/forms/input', () => {
+    // it('should use type attr as CSS class', () => {
+    //     const msg = deep(<Input id="hasIdId" />);
+    //     expect(msg.find('.warning').length).toBe(1);
+    // });
```
Comment (KasparEtter) [inline frontends/web/test/components/forms/input.test.tsx]: Why are these lines commented?

File: frontends/web/test/components/forms/input.test.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    });
+
+    // it('should match the snapshot', () => {
+    //     const msg = deep(
+    //         <Message type="success" style="width:100%">
+    //             <span>hello</span>
+    //         </Message>,
+    //     );
+    //     expect(msg).toMatchSnapshot();
+    // });
```
Comment (KasparEtter) [inline frontends/web/test/components/forms/input.test.tsx]: Why are these lines commented?

File: frontends/web/src/components/password.jsx [hunk 6]
```diff
@@ -74,6 +74,7 @@ export class PasswordSingleInput extends Component {
     }
 
     validate = () => {
+        // @ts-ignore
```
Comment (KasparEtter) [inline frontends/web/src/components/password.jsx]: Why are these @ts-ignore suddenly necessary?

**Pull request merged**